### PR TITLE
api: change API to require all calls be from the same context

### PR DESCRIPTION
**Pull request by jwhui**
This change addresses #76.
This change depends on #75.

>     Change API to require all calls be from the same context.
>     
>     * Remove atomic driver and all uses of otPlatAtomic* in OpenThread.
>     * Change posix example to only use a single thread of execution.

### CODE REVIEW

File: src/ncp/ncp.cpp [hunk 1]
```diff
@@ -146,17 +144,12 @@ Ncp::OutboundFrameSend(void)
     return errorCode;
 }
 
-extern "C" void otPlatSerialSignalSendDone()
-{
-    sSendDoneTask.Post();
-}
-
-void Ncp::SendDoneTask(void *context)
+extern "C" void otPlatSerialSendDone()
```
Comment (pvanhorn) [inline src/ncp/ncp.cpp]: (void)

File: examples/platform/posix/radio.cpp [hunk 2]
```diff
@@ -109,7 +107,7 @@ ThreadError otPlatRadioSetShortAddress(uint16_t address)
     return kThreadError_None;
 }
 
-void hwRadioInit()
+void PlatformRadioInit()
```
Comment (pvanhorn) [inline examples/platform/posix/radio.cpp]: (void)

File: examples/platform/posix/radio.cpp [hunk 3]
```diff
@@ -119,58 +117,44 @@ void hwRadioInit()
 
     s_sockfd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
     bind(s_sockfd, (struct sockaddr *)&sockaddr, sizeof(sockaddr));
-
-    pthread_create(&s_pthread, NULL, &phy_receive_thread, NULL);
 }
 
 ThreadError otPlatRadioEnable()
```
Comment (pvanhorn) [inline examples/platform/posix/radio.cpp]: (void)

File: src/core/mac/mac.cpp [hunk 4]
```diff
@@ -481,23 +478,13 @@ void Mac::HandleBeginTransmit(void)
...
 
-extern "C" void otPlatRadioSignalTransmitDone(void)
+extern "C" void otPlatRadioTransmitDone(bool aRxPending, ThreadError error)
 {
-    sTransmitDoneTask.Post();
+    sMac->TransmitDoneTask(aRxPending, error);
 }
 
-void Mac::TransmitDoneTask(void *aContext)
+void Mac::TransmitDoneTask(bool aRxPending, ThreadError error)
```
Comment (pvanhorn) [inline src/core/mac/mac.cpp]: If you are going to prefix function/method args with "a" then you should do it consistently as in error == aError.

File: src/core/mac/mac.cpp [hunk 5]
```diff
@@ -748,28 +735,21 @@ ThreadError Mac::ProcessReceiveSecurity(const Address &aSrcAddr, Neighbor *aNeig
...
-
-void Mac::ReceiveDoneTask(void *aContext)
+extern "C" void otPlatRadioReceiveDone(ThreadError error)
 {
-    sMac->ReceiveDoneTask();
+    sMac->ReceiveDoneTask(error);
 }
 
-void Mac::ReceiveDoneTask(void)
+void Mac::ReceiveDoneTask(ThreadError aError)
```
Comment (pvanhorn) [inline src/core/mac/mac.cpp]: Here you have aError which I think more closely matches the style throughout.  Just above you have "error".

**Comment by pvanhorn** [comment]
LGTM

File: examples/platform/posix/platform.h [hunk 6]
```diff
@@ -36,34 +36,94 @@
...
+ *
+ */
+void PlatformProcessDrivers(void);
+
+/**
  * This method initializes the alarm service used by OpenThread.
  *
  */
-void hwAlarmInit(void);
+void PlatformAlarmInit(void);
```
Comment (mszczodrak) [inline examples/platform/posix/platform.h]: should AlarmInit be part of platform API? It looks like this is an internal call only, right?

Comment (jwhui) [inline examples/platform/posix/platform.h]: Anything `Platform*` is platform-specific.  In fact, anything not prefixed with `ot*` is platform-specific.

**Comment by darconeous** [comment]
What is the difference between the `Platform*()` namespace and the `otPlat*()` namespace?

**Comment by jwhui** [comment]
The `otPlat*()` namespace are functions that are used to call in/out of OpenThread.  The `Platform*()` namespace are only used in the examples to indicate platform-specific functions that do not interact directly with OpenThread.  Open to other naming suggestions that may help.

**Comment by pvanhorn** [comment]
I was a little confused by the Platform*() naming convention as well. A little digging in the code was necessary to figure it out.  I do prefer it to the older "hw", but it continues to lack the desired intuitiveness. Sadly, I don't have a better option to suggest at this time.

**Pull request merged**